--- Makefile.orig	2025-06-04 12:16:01.056286325 -0400
+++ Makefile	2025-06-04 12:18:47.454879006 -0400
@@ -11,8 +11,7 @@
 
 # Build all of the executable files
 all:
-	$(MAKE) -C src
-	$(MAKE) -C examples
+	$(MAKE) -C src depend libvoro++.a
 
 # Build the help files (with Doxygen)
 help:
@@ -24,16 +23,12 @@
 	$(MAKE) -C examples clean
 
 # Install the executable, man page, and shared library
-install:
-	$(MAKE) -C src
-	$(INSTALL) -d $(IFLAGS_EXEC) $(PREFIX)/bin
+install: all
 	$(INSTALL) -d $(IFLAGS_EXEC) $(PREFIX)/lib
 	$(INSTALL) -d $(IFLAGS_EXEC) $(PREFIX)/man
 	$(INSTALL) -d $(IFLAGS_EXEC) $(PREFIX)/man/man1
 	$(INSTALL) -d $(IFLAGS_EXEC) $(PREFIX)/include
 	$(INSTALL) -d $(IFLAGS_EXEC) $(PREFIX)/include/voro++
-	$(INSTALL) $(IFLAGS_EXEC) src/voro++ $(PREFIX)/bin
-	$(INSTALL) $(IFLAGS) man/voro++.1 $(PREFIX)/man/man1
 	$(INSTALL) $(IFLAGS) src/libvoro++.a $(PREFIX)/lib
 	$(INSTALL) $(IFLAGS) src/voro++.hh $(PREFIX)/include/voro++
 	$(INSTALL) $(IFLAGS) src/c_loops.hh $(PREFIX)/include/voro++
--- src/Makefile.orig	2013-10-17 13:54:13.000000000 -0400
+++ src/Makefile	2025-06-04 12:16:47.293104880 -0400
@@ -10,10 +10,10 @@
 # List of the common source files
 objs=cell.o common.o container.o unitcell.o v_compute.o c_loops.o \
      v_base.o wall.o pre_container.o container_prd.o
-src=$(patsubst %.o,%.cc,$(objs))
+src=$(objs:.o=.cc)
 
 # Makefile rules
-all: libvoro++.a voro++
+all: depend libvoro++.a voro++
 
 depend:
 	$(CXX) -MM $(src) >Makefile.dep
@@ -22,12 +22,12 @@
 
 libvoro++.a: $(objs)
 	rm -f libvoro++.a
-	ar rs libvoro++.a $^
+	$(AR) rs libvoro++.a $(objs)
 
 voro++: libvoro++.a cmd_line.cc
 	$(CXX) $(CFLAGS) -L. -o voro++ cmd_line.cc -lvoro++
 
-%.o: %.cc
+.cc.o:
 	$(CXX) $(CFLAGS) -c $<
 
 help: Doxyfile $(SOURCE)
